### PR TITLE
Move LambdaTerm helpers to classes they belong to

### PR DESCRIPTION
**Pull request by rexim**
I like when the logic is evenly distributed between the entities and not concentrated in a single companion object.

### CODE REVIEW

File: kernel/src/main/scala/me/rexim/morganey/ast/LambdaTerm.scala [hunk 1]
```diff
@@ -82,8 +51,24 @@ case class LambdaFunc(parameter: LambdaVar, body: LambdaTerm) extends LambdaTerm
 
   override val freeVars: Set[String] = body.freeVars - parameter.name
 
+  private[ast] def nestedFunctions(): (List[LambdaVar], LambdaTerm) = {
+    @tailrec
+    def go(func: LambdaFunc, acc: List[LambdaVar]): (List[LambdaVar], LambdaTerm) =
```
Comment (keddelzz) [inline kernel/src/main/scala/me/rexim/morganey/ast/LambdaTerm.scala:56]: This change will throw a `MatchError` at runtime, if this method was called from an instance with the wrong type. This method is only allowed to be called on instances of `LambdaFunc`.

File: kernel/src/main/scala/me/rexim/morganey/ast/LambdaTerm.scala [hunk 2]
```diff
@@ -99,12 +84,22 @@ case class LambdaApp(leftTerm: LambdaTerm, rightTerm: LambdaTerm) extends Lambda
 
   override val freeVars: Set[String] = leftTerm.freeVars ++ rightTerm.freeVars
 
+  private[ast] def nestedApplications(): (LambdaTerm, List[LambdaTerm]) = {
+    @tailrec
+    def go(app: LambdaApp, acc: List[LambdaTerm]): (LambdaTerm, List[LambdaTerm]) =
+      app match {
+        case LambdaApp(l: LambdaApp, r) => go(l, r :: acc)
```
Comment (keddelzz) [inline kernel/src/main/scala/me/rexim/morganey/ast/LambdaTerm.scala:91]: Same thing as with [167-comment](https://github.com/rexim/Morganey/pull/167#discussion-diff-77370000), but the "right" type here is `LambdaApp`.

File: kernel/src/main/scala/me/rexim/morganey/ast/LambdaTerm.scala (hunk 1 above)
Comment (rexim) [reply kernel/src/main/scala/me/rexim/morganey/ast/LambdaTerm.scala:56]: That's why I made `nestedFunctions` a member of `LambdaFunc`. It is not possible to invoke that `go` with something that is not `LambdaFunc`. Please correct me if I'm wrong.

Comment (keddelzz) [reply kernel/src/main/scala/me/rexim/morganey/ast/LambdaTerm.scala:56]: You are right, my mistake ;) I should have pulled your branch and double checked locally before complaining :)

Comment (rexim) [reply kernel/src/main/scala/me/rexim/morganey/ast/LambdaTerm.scala:56]: No-no! Please complain about my code! You made me double check it, which is good. :)

File: kernel/src/main/scala/me/rexim/morganey/ast/LambdaTerm.scala (hunk 2 above)
Comment (rexim) [reply kernel/src/main/scala/me/rexim/morganey/ast/LambdaTerm.scala:91]: It is member of LambdaApp, so I guess it's the same situation as https://github.com/rexim/Morganey/pull/167#discussion_r77370535

Comment (keddelzz) [reply kernel/src/main/scala/me/rexim/morganey/ast/LambdaTerm.scala:91]: > so I guess it's the same situation as

Yes :)

**Comment by keddelzz** [comment]
LGTM :+1:

**Comment by rexim** [comment]
@keddelzz cool! Thank you very much for the review!

**Pull request merged**